### PR TITLE
fix: allow rowid in columns for scanner

### PR DESCRIPTION
**Pull request by Jay-ju**
Currently, the scanner does not support passing in hidden columns, such as rowid and rowaddr, which will result in an error.
```
self = <lance.dataset.LanceScanner object at 0x7f330d7cbf60>

    def to_reader(self) -> pa.RecordBatchReader:
>       return self._scanner.to_pyarrow()
E       ValueError: LanceError(Schema): Schema error: No field named _rowid. Valid fields are data_item_id., /data00/code/emr/lance/rust/lance-datafusion/src/projection.rs:171:25

```

This PR mainly allows both with_row_id in the scanner and rowid in the columns. The reason for this is that when performing push - down on the computing side, the schema in the dataframe is generally pushed directly to the columns, rather than setting properties specifically for a certain with_row_id or with_row_addr. 

Therefore, there is a hope to make a compatibility on the lance side, similar to allowing such behavior. 
```
lance_ds = lance.dataset(path)
table = lance_ds.scanner(
        with_row_id=True, columns=["data_item_id", "_rowid"]
    ).to_table()
```

### CODE REVIEW

**Comment by LuQQiu** [comment]
@wjones127 are we planning to expose row_id and row_address to end users? will they subject to change?

**Comment by Jay-ju** [comment]
> @wjones127 are we planning to expose row_id and row_address to end users? will they subject to change?

@LuQQiu At the current stage, as I understand it, it can only be disclosed. It may not be necessary in the future (if there is a business key), because the screening of a lot of pre-training data is based on the set of row_id to identify which data needs to be selected for different datasets.

**Comment by westonpace** [comment]
> @wjones127 are we planning to expose row_id and row_address to end users? will they subject to change?

We haven't exposed them in `lancedb` but in `lance` we expose both and assume the user is advanced enough to understand the risks.

File: rust/lance-core/src/datatypes/schema.rs [hunk 1]
```diff
@@ -1081,8 +1085,9 @@ impl Projection {
 
     /// Convert the projection into an Arrow schema
     pub fn to_arrow_schema(&self) -> Result<arrow_schema::Schema> {
-        let mut arrow_schema: arrow_schema::Schema = (&self.to_schema()).into();
+        let arrow_schema: arrow_schema::Schema = (&self.to_schema()).into();
         // Should we be adding row_id / row_addr on to_schema?
+        let mut arrow_schema = arrow_schema.clone();
```
Comment (westonpace) [inline rust/lance-core/src/datatypes/schema.rs]: What is this change for?

**Comment by Jay-ju** [comment]
> I think this only works if `with_row_id` is also passed. For example, this would fail...
> 
> ```
> lance_ds.scanner(columns=["data_item_id", "_rowid"]).to_table()
> ```
> 
> When we see the column named `_rowid` do we want to automatically set `with_row_id` to true? (same with `_rowaddr`)


@westonpace Well, there should indeed be such a restriction.

File: rust/lance-core/src/datatypes/schema.rs (hunk 1 above)
Comment (Jay-ju) [reply rust/lance-core/src/datatypes/schema.rs]: It has been deleted here

**Comment by Jay-ju** [comment]
@westonpace Could you help review it when you have time?

**Comment by Jay-ju** [comment]
@westonpace @jackye1995 Do you have any questions regarding this PR? Could you please take a look at the changes here? Thank you first

File: python/python/tests/test_lance.py [hunk 2]
```diff
@@ -248,3 +248,22 @@ def test_io_counters(tmp_path):
...
+            "data_item_id": pa.array([1001, 1002, 1003]),
+            "a": pa.array([1, 2, 3]),
+        }
+    )
+
+    lance.write_dataset(data, path)
+
+    lance_ds = lance.dataset(path)
+    table = lance_ds.scanner(
+        with_row_id=True, columns=["data_item_id", "_rowid"]
```
Comment (jackye1995) [inline python/python/tests/test_lance.py]: I think we should also test a negative case of not setting with_row_id and expect an error

File: python/python/tests/test_lance.py [hunk 3]
```diff
@@ -248,3 +248,22 @@ def test_io_counters(tmp_path):
     dataset.to_table()
     assert lance.iops_counter() > starting_iops
     assert lance.bytes_read_counter() > starting_bytes
+
+
+def test_with_row_id(tmp_path):
```
Comment (jackye1995) [inline python/python/tests/test_lance.py]: we should also have a test of `test_with_row_addr`

File: python/src/dataset.rs [hunk 4]
```diff
@@ -665,14 +674,6 @@ impl Dataset {
 
         scanner.scan_in_order(scan_in_order.unwrap_or(true));
 
-        if with_row_id.unwrap_or(false) {
```
Comment (jackye1995) [inline python/src/dataset.rs:679]: this change is just moving the code up? Does it serve any functional purpose?

File: rust/lance/src/dataset/scanner.rs [hunk 5]
```diff
@@ -474,9 +474,19 @@ impl Scanner {
         &mut self,
         columns: &[(impl AsRef<str>, impl AsRef<str>)],
     ) -> Result<&mut Self> {
+        let filtered_columns: Vec<_> = columns
```
Comment (jackye1995) [inline rust/lance/src/dataset/scanner.rs:502]: we should have a test in `scanner.rs` to verify this behavior

Comment (Jay-ju) [reply rust/lance/src/dataset/scanner.rs:502]: @jackye1995  Some test cases in rs have been added here

File: python/python/tests/test_lance.py (hunk 3 above)
Comment (Jay-ju) [reply python/python/tests/test_lance.py]: add it

File: python/python/tests/test_lance.py (hunk 2 above)
Comment (Jay-ju) [reply python/python/tests/test_lance.py]: add it

File: python/src/dataset.rs (hunk 4 above)
Comment (Jay-ju) [reply python/src/dataset.rs:679]: Yes, it does work. Originally, the retrieval of with_row_id was after the project. Now it has been moved up here.

```python
(Some(c), None) => {
                scanner
                    .project(&c)
                    .map_err(|err| PyValueError::new_err(err.to_string()))?;
            }
            (None, Some(ct)) => {
                scanner
                    .project_with_transform(&ct)
                    .map_err(|err| PyValueError::new_err(err.to_string()))?;
            }
``` 

In front of this code, if `with_row_id=True`, some `_rowid` will be automatically added to the project.


@jackye1995 Please help review it when you have time. It has been modified according to the comment.

File: rust/lance/src/dataset/scanner.rs [hunk 6]
```diff
@@ -6393,4 +6431,78 @@ mod test {
...
+        let mut scanner = dataset.scan();
+        scanner.with_row_id = false;
+        scanner.with_row_address = false;
+        let result = scanner.project(&[if with_row_id { ROW_ID } else { ROW_ADDR }]);
+        assert!(result.is_err());
+    }
+
+    #[rstest]
+    #[tokio::test]
+    async fn test_with_row_id() {
```
Comment (jackye1995) [inline rust/lance/src/dataset/scanner.rs]: nit: it feels unnecessary to test this independently with a helper function. Can we just have one test `test_with_row_id_and_address` that test the 3 cases in 1 pass?

File: python/python/tests/test_lance.py [hunk 7]
```diff
@@ -248,3 +248,34 @@ def test_io_counters(tmp_path):
     dataset.to_table()
     assert lance.iops_counter() > starting_iops
     assert lance.bytes_read_counter() > starting_bytes
+
+
+@pytest.mark.parametrize(
+    "row_param, column_name",
+    [("with_row_id", "_rowid"), ("with_row_address", "_rowaddr")],
+)
+def test_with_row_id(tmp_path, row_param, column_name):
```
Comment (jackye1995) [inline python/python/tests/test_lance.py:257]: nit: `test_with_row_id_and_address`, given we are now testing both

File: rust/lance/src/dataset/scanner.rs (hunk 6 above)
Comment (Jay-ju) [reply rust/lance/src/dataset/scanner.rs]: fix it

File: python/python/tests/test_lance.py (hunk 7 above)
Comment (Jay-ju) [reply python/python/tests/test_lance.py:257]: add it

Comment (Jay-ju) [reply python/python/tests/test_lance.py:257]: @jackye1995 Please help take a look when you have time. I've been a bit busy this week, so I'm a bit late. Sorry about that.

**Review by jackye1995 — APPROVED** [review]
looks good to me

**Comment by jackye1995** [comment]
CI failure is unrelated, merging.
